pkg/envtest/webhook.go

```diff
@@ -0,0 +1,358 @@
+	// it will be automatically populated
+	LocalServingHost string
+
+	// LocalServingPort is the allocated port for serving webhooks on.
+	// it will be automatically populated by a random available local port
+	LocalServingPort int
+
+	// LocalServingCertDir is the allocated directory for serving certificates.
+	// it will be automatically populated by the local temp dir
+	LocalServingCertDir string
+
+	// MaxTime is the max time to wait
+	MaxTime time.Duration
+
+	// PollInterval is the interval to check
+	PollInterval time.Duration
+}
+
+// ModifyWebhookDefinitions modifies webhook definitions by:
+// - applying CABundle based on the provided tinyca
+// - if webhook client config uses service spec, it's removed and replaced with direct url
+func (o *WebhookInstallOptions) ModifyWebhookDefinitions(caData []byte) error {
+	hostPort, err := o.generateHostPort()
+	if err != nil {
+		return err
+	}
+
+	for _, hookSet := range o.MutatingWebhooks {
+		for i := range hookSet.Webhooks {
+			cfg := &hookSet.Webhooks[i].ClientConfig
+			cfg.CABundle = caData
+			var path string
+			if cfg.Service != nil && cfg.Service.Path != nil {
+				path = *cfg.Service.Path
+			}
+			cfg.Service = nil // we cannot use service in integration tests since we're running controller outside cluster
+			url := fmt.Sprintf("https://%s/%s", hostPort, path)
+			cfg.URL = &url
+		}
+	}
+	for _, hookSet := range o.ValidatingWebhooks {
+		for i := range hookSet.Webhooks {
+			cfg := &hookSet.Webhooks[i].ClientConfig
+			cfg.CABundle = caData
+			var path string
+			if cfg.Service != nil && cfg.Service.Path != nil {
+				path = *cfg.Service.Path
+			}
+			cfg.Service = nil // we cannot use service in integration tests since we're running controller outside cluster
+			url := fmt.Sprintf("https://%s/%s", hostPort, path)
+			cfg.URL = &url
+		}
+	}
+	return nil
+}
+
+func (o *WebhookInstallOptions) generateHostPort() (string, error) {
+	port, host, err := addr.Suggest()
+	if err != nil {
+		return "", fmt.Errorf("unable to grab random port for serving webhooks on: %v", err)
+	}
+	o.LocalServingPort = port
+	o.LocalServingHost = host
+	return net.JoinHostPort(host, fmt.Sprintf("%d", port)), nil
+}
+
+// Install installs specified webhooks to the API server
+func (o *WebhookInstallOptions) Install(config *rest.Config) error {
+	hookCA, err := o.setupCA()
+	if err != nil {
+		return err
+	}
+	if err := parseWebhookDirs(o); err != nil {
+		return err
+	}
+
+	err = o.ModifyWebhookDefinitions(hookCA)
+	if err != nil {
+		return err
+	}
+
+	if err := createWebhooks(config, o.MutatingWebhooks, o.ValidatingWebhooks); err != nil {
+		return err
+	}
+
+	if err := WaitForWebhooks(config, o.ValidatingWebhooks, o.MutatingWebhooks, *o); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// WaitForWebhooks waits for the Webhooks to be available through API server
+func WaitForWebhooks(config *rest.Config,
+	validatingHooks []*admissionreg.ValidatingWebhookConfiguration,
```

Comment:
Not excited about this API. It only handles v1beta1 webhooks.
It will be a breaking change later, if we want to support v1.
It will be great if it takes a list of objects and then handle them differently depends on the version and kind.

Reply:
makes total sense, thanks